keymgmt/keymgmt.c

```diff
@@ -23,11 +23,74 @@ usage(void)
 	    "--outkeyfile new-key-file", "[--passphrased]",
 	    "[--passphrase-mem maxmem]", "[--passphrase-time maxtime]",
 	    "[-r] [-w] [-d] [--nuke]");
+	fprintf(stderr, "       tarsnap-keymgmt --print-key-id key-file\n");
+	fprintf(stderr, "       tarsnap-keymgmt --print-key-permissions "
+	    "key-file\n");
 	exit(1);
 
 	/* NOTREACHED */
 }
 
+static void
+print_id(const char *keyfilename)
+{
+	uint64_t machinenum = (uint64_t)(-1);
+
+	/* Read keyfile and machine name. */
+	if (keyfile_read(keyfilename, &machinenum, ~0)) {
+		warnp("Cannot read key file: %s", keyfilename);
+		exit(1);
+	}
+
+	/* Print key ID. */
+	fprintf(stdout, "%li\n", machinenum);
+	exit(0);
+
+	/* NOTREACHED */
+}
+	
+static void
+print_permissions(const char *keyfilename)
+{
+	uint64_t machinenum = (uint64_t)(-1);
+	int has_read;
+	int has_write;
+	int has_delete;
+
+	/* Read keyfile and machine name. */
+	if (keyfile_read(keyfilename, &machinenum, ~0)) {
+		warnp("Cannot read key file: %s", keyfilename);
+		exit(1);
+	}
+
+	/* Determine permissions. */
+	has_read = (crypto_keys_missing(CRYPTO_KEYMASK_READ) == NULL);
+	has_write = (crypto_keys_missing(CRYPTO_KEYMASK_WRITE) == NULL);
+	has_delete = (crypto_keys_missing(CRYPTO_KEYMASK_AUTH_DELETE) == NULL);
+
+	/* Print key permissions. */
+	fprintf(stdout, "This key has permissions for: ");
+	if (has_read && has_write && has_delete)
+		fprintf(stdout, "reading, writing, and deleting.\n");
+	if (has_read && has_write && !has_delete)
+		fprintf(stdout, "reading and writing.\n");
+	if (has_read && !has_write && has_delete)
+		fprintf(stdout, "reading and deleting.\n");
+	if (has_read && !has_write && !has_delete)
+		fprintf(stdout, "reading.\n");
+	if (!has_read && has_write && has_delete)
+		fprintf(stdout, "writing and deleting.\n");
```

Comment:
Make this "writing and nuking", since you can't delete archives without being able to read them.